# Map Reads

## Learning objectives

- Download reference genome and annotations
- Map reads to reference genome with STAR
- Generate index for BAM files
- Generate QC report for mapped reads

## Purpose

Mapping high-quality reads to a reference is a fundamental step in a typical RNA-seq analysis.

This mapping step assigns a read to the genomic locus that it is (likely) derived from.
We can use this mapping information to generate a variety of data for downstream analysis such as determining the expression levels of genes, identifying novel splice sites, and identifying genomic variants[^1].

After read trimming, we run FastQC on the trimmed reads to make sure the trimming mitigated any issues observed at the raw reads QC step.

[^1]: The types of analyses we can perform are determined by factors such as availability of genomic resources, library construction, sequencing depth, etc.

## Overview of mapping workflow

```{mermaid}
flowchart LR
    ref_genome[(Reference\nGenome)]
    ref_annotations[(Reference\nAnnotations)]
    build_index{STAR\ngenomeGenerate}
    ref_index[(Reference\nIndex)]

    ref_genome --- build_index
    ref_annotations --- build_index
    build_index --> ref_index
```

## Software usage

### STAR usage

#### Software availability

STAR is available as a module that can be loaded on ISAAC:

``` {.bash filename="Terminal"}
module load star/2.7.6a
```


Alternatively, more recent versions are available as Apptainer containers through BioContainers:

```{.bash filename="Terminal"}
function STAR {
  local image_uri="https://depot.galaxyproject.org/singularity/star%3A2.7.11b--h43eeafb_1"
  apptainer exec "${image_uri}" STAR "${@}"
}
```


Both of the methods above make STAR available through the `STAR` command.


### Command usage --- Generate genome index

Basic usage:

``` {.bash filename="Terminal"}
STAR \
  --runMode genomeGenerate \
  --genomeFastaFiles genome.fasta \
  --sjdbGTFfile annotations.gtf \
  --genomeDir STAR
```

